RationaleMCP/0027/HindleyMilner.md

```diff
@@ -0,0 +1,195 @@
+# Unit checking
+Unit checking is performed by using a type inference algorithm, albeit adapted to work with units. An example of such an algorithm is the Hindley-Milner algorithm.
+At a high level the algorithm collects a set of constraints on unit of expressions. It extracts a sequence of substitutions from these constraints which once applied constructs the unit of a given expression.
+Any inconsistencies discovered during this process, is a unit error, and any unit expression that can't be evaluated is considered undefined.
+
+## Unit of expressions
+
+The unit of an arbitrary Modelica expression is represented by a unit meta-expression.
+This sections outlines the different constructs it comprises.
+
+### Variables
+The unit of variable `var`, is represented by identifier `var.unit`.
+
+### Literals
+
+#### `Well-formed` unit
```

Comment:
Here's a somewhat bigger discussion that I'm randomly attaching to this line.

It seems to me that we're missing a name for those units which are neither empty nor undefined, and I'm thinking that _well-formed unit_ could be a suitable name for that?

With such concepts, instead of saying that empty and undefined units are literals, couldn't one say that these belong to the alternatives for a _meta unit_, defined as being either of:
- empty unit
- undefined unit
- well-formed unit

One could then state (and later prove) that unit variables (constants aside) can only ever be assigned well-formed units during the process.

What is currently called a well-formed unit (literal) could then be a _concrete unit_ (which would be the only kind of literal now that the others are just alternatives for a _meta unit_), while _well-formed unit_ could be defined recursively as either of:
- concrete unit (avoiding the term _unit literal_ due to possible confusion with the unit expression string syntax)
- unit variable (unassigned or assigned to a well-formed unit)
- product of well-formed units
- well-formed unit raised to a rational power
- time derivative of well-formed unit

Finally, the goal would be to define the operators below as mappings from meta units to meta units.  If this succeeds, it will be clear that empty and undefined units are not lurking deep inside meta units.

Reply:
This has now been discussed.  Resolving.